app/src/main/java/com/poupa/vinylmusicplayer/service/MusicService.java

```diff
@@ -645,186 +647,155 @@ public Song getCurrentSong() {
-                if (isLastTrack()) {
-                    position = 0;
-                }
-                break;
-            case REPEAT_MODE_THIS:
-                if (force) {
-                    if (isLastTrack()) {
-                        position = 0;
-                    }
-                } else {
-                    position -= 1;
-                }
-                break;
-            default:
-            case REPEAT_MODE_NONE:
-                if (isLastTrack()) {
-                    position -= 1;
-                }
-                break;
+    public IndexedSong getIndexedSongAt(int position) {
+        if (position >= 0 && position < playingQueue.size()) {
+            return playingQueue.getIndexedSongAt(position);
+        } else {
+            return new IndexedSong(Song.EMPTY_SONG, -1);
         }
-        return position;
     }
 
     public boolean isLastTrack() {
-        return getPosition() == getPlayingQueue().size() - 1;
+        return playingQueue.isLastTrack();
     }
 
     public ArrayList<Song> getPlayingQueue() {
-        return playingQueue;
+        return playingQueue.getPlayingQueueSongOnly();
     }
 
     public int getRepeatMode() {
-        return repeatMode;
+        return playingQueue.getRepeatMode();
     }
 
     public void setRepeatMode(final int repeatMode) {
-        switch (repeatMode) {
-            case REPEAT_MODE_NONE:
-            case REPEAT_MODE_ALL:
-            case REPEAT_MODE_THIS:
-                this.repeatMode = repeatMode;
-                PreferenceManager.getDefaultSharedPreferences(this).edit()
-                        .putInt(SAVED_REPEAT_MODE, repeatMode)
-                        .apply();
-                prepareNext();
-                handleAndSendChangeInternal(REPEAT_MODE_CHANGED);
-                break;
-        }
+        playingQueue.setRepeatMode(repeatMode);
+
+        PreferenceManager.getDefaultSharedPreferences(this).edit()
+                .putInt(SAVED_REPEAT_MODE, repeatMode)
+                .apply();
+        prepareNext();
+        handleAndSendChangeInternal(REPEAT_MODE_CHANGED);
     }
 
-    public void openQueue(@Nullable final ArrayList<Song> playingQueue, final int startPosition, final boolean startPlaying) {
-        if (playingQueue != null && !playingQueue.isEmpty() && startPosition >= 0 && startPosition < playingQueue.size()) {
-            // it is important to copy the playing queue here first as we might add/remove songs later
-            originalPlayingQueue = new ArrayList<>(playingQueue);
-            this.playingQueue = new ArrayList<>(originalPlayingQueue);
-
-            int position = startPosition;
-            if (shuffleMode == SHUFFLE_MODE_SHUFFLE) {
-                ShuffleHelper.makeShuffleList(this.playingQueue, startPosition);
-                position = 0;
-            }
+    public void cycleRepeatMode() {
+        playingQueue.cycleRepeatMode();
+    }
+
+    private void shuffleChangeSignaling() {
```

Comment:
Nitpicking: I'd rename this to `propagateShuffleChange`

Reply:
ok